### ISSUE TITLE
Oc/Dc2Pointer no-hash hash

### ISSUE THREAD

**Issue by orenshatz** [knossos-project/knossos#225]
As long the supercube is a rectangular cube, the current unordered_map is a very inefficient solution for a quick retrieval of a cube from coordinate in the supercube. A simple 3D table of the supercube would be the best - dereference would cost practically nothing.
It would only have to be copied - also, very quickly - when the supercube is changing.

It seems that somehow the current unordered_map hashing is not a bottleneck in normal knossos usage, including high-throughput random voxel-wise access from python, probably due to a heavier burden on the actual thunking. However, should in the future more efficient random access schemes be used within the C++ code, that cannot be efficiently binned into cubes to avoid pre-hashing the cube pointer for each bin - a hashed access to cubes may be the bottleneck, and direct table reference the solution.

**Comment by Optiligence** [comment]
Maybe we want to change the supercube shape soonish…
I don’t think this is needed, one should not query cubes that often. 
As you said, it’s no bootleneck.

**Comment by orenshatz** [comment]
As you know I already suggested a rectangular one soon, and an arbitrary one later, but just in case we don't go for it...
I'm not sure what you mean by "should", the user can decide whatever he wants, and a coordinate-list (vector in fact) overlay access would soon be introduced to the python interface.
The only question is whether we want to enforce binning on plugin developers who want high performance, or we make life easier for them.

**Comment by Optiligence** [comment]
If you query cubes that often, you’re doing something wrong (and are obviously not concerned about performance).

**Comment by orenshatz** [comment]
No, if you query cubes so often, you must be a user who's random-accessing voxels and doesn't even care what a cube is - which is just ideal.

**Comment by Optiligence** [comment]
then you don’t query cubes but
> a coordinate-list (vector in fact) overlay access

is voxel wise and cube-aware
it could order the coordinates by cube if they appear to be truly random and it’s really a performance problem
however there’s still no need for the no-hash

**Comment by orenshatz** [comment]
Cube-aware and therefore costly.
As I've said and you agree, binning is expensive.
The no-hash would render it unnecessary.
In fact the no-hash is super easy to implement and has no disadvantage whatsoever.
I will implement it when I have time.

**Comment by Optiligence** [comment]
>Cube-aware and therefore costly.

Wut? 

>As I've said and you agree, binning is expensive.

I don’t agree that it’s expensive enough to arise the need to optimize. In fact there is nothing to suggest that. 

>In fact the no-hash is super easy to implement and has no disadvantage whatsoever.

Unnecessary/needless code bloat. 

>I will implement it when I have time.

I want to relief you of that adventure through this discussion, but if you really feel the need to.

**Comment by orenshatz** [comment]
- Cube-aware meaning you have to hash for each coordinate.
- It's expensive as a hash is computationally more expensive than anything which is not an algorithm.
In fact per-voxel a hash is probably one of the most expensive algorithm in the world today. Therefore it's definitely not scalable for random access on e.g. M=11.
- It's actually the other way around, the new implementation would probably be much lighter.
- I want to relief ourselves from future profiling, as the direct table approach would never be the bottleneck, so far we couldn't say the same of the hash, I actually had to spend time checking it, and it most likely would be the same in the future.

**Comment by orenshatz** [comment]
(Mind you everything we say also applies to Dc, not only Oc.
So the fact the existing overlay algorithms may be more expensive than a hash
shouldn't mislead you that the same is true for Dc.)

**Comment by Optiligence** [comment]
>Cube-aware meaning you have to hash for each coordinate.
>Therefore it's definitely not scalable for random access on e.g. M=11.

There are only so many cubes especially for a huge amount of voxels. 

>It's expensive as a hash is computationally more expensive than anything which is not an algorithm. 

measure it, it’s not much

>In fact per-voxel a hash is probably one of the most expensive algorithms in the world today. 

Well we wont hash per voxel anyway. 

>It's actually the other way around, the new implementation would probably be much lighter.

It cannot replace the current one. 

>I want to relief ourselves from future profiling

Without profiling there’s no need.

**Comment by orenshatz** [comment]
- You're ignoring the point, it's not about the number of cubes, it's whether we must bin batch arbitrary voxel-wise operations into cubes in order to carry them out efficiently. We don't. It forces a minimum granularity on the size of the batch.
- It is at least several (implementation specific) arithmetic operations for the hash value alone, then collision handling (which should not apply in our case, as the hash space is considerably bigger than the key space).
- Well we will, if we don't bin.
- Actually it can, quite easy.
- I already had to profile today to rule out the hash as the bottleneck. I would have to do it again. If your argument is that the hash would never be a bottleneck since we will never profile it, then we might as well say there is no bottleneck at all, as we don't need to profile anything at all.

**Comment by Optiligence** [comment]
>It forces a minimum granularity on the size of the batch.
>Well we will, if we don't bin.

Yes and i think that’s okay. If you want performance you should do batch operations, if you don’t its irrelevant. Calling a function (from python) for every voxel is the mad thing here, not the hash lookup inside. 

>Actually it can, quite easy.

Not without preventing currently available possibilities. 

>If your argument is that the hash would never be a bottleneck since we will never profile it, then we might as well say there is no bottleneck at all, as we don't need to profile anything at all.

What?

**Comment by orenshatz** [comment]
- No, *you* decided that *you* implement performance by batch operations. But any other programmer or python user might think otherwise, and *your* decision should not bind him if we have an alternative. The concept of random data access is quite established. The hash itself is an attempt to achieve such access, it's simply not meeting top performance demands. Calling something mad doesn't make it mad, and as I made clear, a single python invocation with a long coordinate-list already presents a problem. Binning may be one way to solve this specific problem, but then you introduce more code for the purpose of solving this specific problem, than the actual code it would take to simply replace the hash. In addition, other random access usages might be more dynamic and not easily batched. The entire argument is flawed, the only reason why something should *not* be done, is not that it is not needed according to *your* opinion, but why it makes thing worse.
- Which current possibilities are these? If you refer to arbitrary shaped supercube, it's not current.
- You wrote "Without profiling there’s no need. ". And how exactly do you decide whether you profile something?

Anyway, this entire argument would be needless if the actual hash function of the gcc 4.9 implementation was well documented, instead of stuck somewhere in a pile of code.

**Comment by orenshatz** [comment]
Hash function alone (no data structure handling) performs 20M/sec which is very good.
Since most voxel-wise operations are spatially correlated, an effective no-hash can be achieved
according to Joergen's suggesting by simply keeping the last queried Coordinate-Cube pair as cache,
and invalidating it whenever the hash is changed (i.e. element entered or removed).
